### PR TITLE
Add pre-commit

### PR DESCRIPTION
**Pull request by realloc**
Adding [pre-commit](https://pre-commit.com/) framework support.

It covers our basic needs, but there is a room for improvement:
- Add GitHub issue existence check
- Add version match check for tag/VERSION/deb

### CODE REVIEW

**Review by acid-ant — APPROVED** [review]

**Comment by alexvanin** [comment]
Is it possible to test pre-commit.ci in this PR?

**Comment by realloc** [comment]
> Is it possible to test pre-commit.ci in this PR?

Hope so, working on this.

**Comment by realloc** [comment]
> Is it possible to test pre-commit.ci in this PR?

pre-commit.ci doesn't add nice sections in Checks tab. Trying separate action.

If both ways fail, we can add check status with external Jenkins, as we did some time ago.

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    rev: v4.3.0
+    hooks:
+    - id: check-added-large-files
+    - id: check-case-conflict
+    - id: check-executables-have-shebangs
+    - id: check-shebang-scripts-are-executable
+    - id: check-merge-conflict
+    - id: check-json
+    - id: check-xml
+    - id: check-yaml
```
Comment (fyrchik) [inline .pre-commit-config.yaml:12]: If any of this fails, can we see all errors or only the first?

File: Makefile [hunk 2]
```diff
@@ -132,6 +132,14 @@ test:
 lint:
 	@golangci-lint --timeout=5m run
 
+# Activate pre-commit hooks
+pre-commit:
+	pre-commit install -t pre-commit -t commit-msg
```
Comment (fyrchik) [inline Makefile:137]: If I change a commit message during rebase, will all hooks be executed on only message check?

File: .gitlint [hunk 3]
```diff
@@ -0,0 +1,10 @@
+[general]
+fail-without-commits=true
+contrib=CC1
+
+[title-match-regex]
+regex=^\[\#[0-9]*\]\s
```
Comment (fyrchik) [inline .gitlint:6]: `^\[\#[0-9]+\]\s.+$`?

File: .github/workflows/pre-commit.yml [hunk 4]
```diff
@@ -0,0 +1,17 @@
+name: pre-commit
+
+on:
+  pull_request:
+  push:
+    branches: [main]
```
Comment (fyrchik) [inline .github/workflows/pre-commit.yml:6]: `master`?
Can we use regex for support?

**Comment by fyrchik** [comment]
Also, can we globally exclude autogenerated files from most of the checks? I can imagine a situation where a linter will complain.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (realloc) [reply .pre-commit-config.yaml:12]: Should see all errors. Will check again.

File: Makefile (hunk 2 above)
Comment (realloc) [reply Makefile:137]: No, it should not.

**Comment by realloc** [comment]
> Also, can we globally exclude autogenerated files from most of the checks? I can imagine a situation where a linter will complain.

We can. For the `neofs-node`, what do you propose to exclude?

**Comment by realloc** [comment]
@alexvanin Let's keep precommit.ci for now. It looks like a normal GH Actions check. We can do all cosmetically fancy stuff later with the dedicated Jenkins.

**Comment by fyrchik** [comment]
@realloc at least all `**/*.pb.go` files.

**Review by fyrchik — APPROVED** [review]

**Pull request closed without merging**